bundles/org.eclipse.equinox.p2.director/src/org/eclipse/equinox/internal/p2/director/Slicer.java

```diff
@@ -31,42 +31,42 @@ public class Slicer {
 	private final IQueryable<IInstallableUnit> possibilites;
 	private final boolean considerMetaRequirements;
 	protected final IInstallableUnit selectionContext;
-	private final Map<String, Map<Version, IInstallableUnit>> slice; //The IUs that have been considered to be part of the problem
-	private final MultiStatus result;
+	/** The IUs that have been considered to be part of the problem */
+	private final Map<String, Map<Version, IInstallableUnit>> slice = new HashMap<>();
+	private final MultiStatus result = new MultiStatus(Slicer.class, 0, Messages.Planner_Problems_resolving_plan);
 
-	private LinkedList<IInstallableUnit> toProcess;
-	private Set<IInstallableUnit> considered; //IUs to add to the slice
-	private Set<IInstallableUnit> nonGreedyIUs = new HashSet<>(); //IUs that are brought in by non greedy dependencies
+	private Queue<IInstallableUnit> toProcess;
+	private Set<IInstallableUnit> considered; // IUs to add to the slice
+	private final Set<IInstallableUnit> nonGreedyIUs = new HashSet<>(); // IUs that are brought in by non greedy dependencies
 
 	public Slicer(IQueryable<IInstallableUnit> input, Map<String, String> context, boolean considerMetaRequirements) {
 		this(input, InstallableUnit.contextIU(context), considerMetaRequirements);
 	}
 
-	public Slicer(IQueryable<IInstallableUnit> possibilites, IInstallableUnit selectionContext, boolean considerMetaRequirements) {
+	public Slicer(IQueryable<IInstallableUnit> possibilites, IInstallableUnit selectionContext,
+			boolean considerMetaRequirements) {
 		this.possibilites = possibilites;
 		this.selectionContext = selectionContext;
 		this.considerMetaRequirements = considerMetaRequirements;
-		slice = new HashMap<>();
-		result = new MultiStatus(DirectorActivator.PI_DIRECTOR, IStatus.OK, Messages.Planner_Problems_resolving_plan, null);
 	}
 
-	public IQueryable<IInstallableUnit> slice(IInstallableUnit[] ius, IProgressMonitor monitor) {
+	public IQueryable<IInstallableUnit> slice(Collection<IInstallableUnit> ius, IProgressMonitor monitor) {
```

Comment:
This seem to have broken the i-build:
https://ci.eclipse.org/releng/job/Builds/job/I-build-4.30/78/console

> 00:16:41  [ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:4.0.4-SNAPSHOT:compile (default-compile) on project org.eclipse.pde.core: Compilation failure: Compilation failure: 
> 00:16:41  [ERROR] /home/jenkins/agent/workspace/Builds/I-build-4.30/eclipse.platform.releng.aggregator/eclipse.platform.releng.aggregator/cje-production/gitCache/eclipse.platform.releng.aggregator/eclipse.pde/ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/target/IUBundleContainer.java:[321] 
> 00:16:41  [ERROR] 	IQueryable<IInstallableUnit> slice = slicer.slice(fUnits, new NullProgressMonitor());
> 00:16:41  [ERROR] 	                                            ^^^^^
> 00:16:41  [ERROR] The method slice(Collection<IInstallableUnit>, IProgressMonitor) in the type Slicer is not applicable for the arguments (IInstallableUnit[], NullProgressMonitor)
> 00:16:41  [ERROR] /home/jenkins/agent/workspace/Builds/I-build-4.30/eclipse.platform.releng.aggregator/eclipse.platform.releng.aggregator/cje-production/gitCache/eclipse.platform.releng.aggregator/eclipse.pde/ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/target/P2TargetUtils.java:[1364] 
> 00:16:41  [ERROR] 	IQueryable<IInstallableUnit> slice = slicer.slice(units, subMonitor.split(50));
> 00:16:41  [ERROR] 	                                            ^^^^^
> 00:16:41  [ERROR] The method slice(Collection<IInstallableUnit>, IProgressMonitor) in the type Slicer is not applicable for the arguments (IInstallableUnit[], SubMonitor)
> 00:16:41  [ERROR] 2 problems (2 errors)
> 00:16:41  [ERROR] -> [Help 1]

Reply:
This is probably unrelated to **this** PR but I think changing the signature of the `slice` method also broke the installation process (Oomph) so I'm commenting here anyway (if you know a better place, please let me know):
I just installed a new Eclipse with the Oomph-Installer (I'm using Mac) and when it got to the point of **executing manual tasks** after Eclipse opened for the first time, I got this error when computing the step **Modular Target**:

```java
Performing Targlets Modular Target (eclipse-sdk-prereqs + ECF Dependency + EMF Dependency + Eclipse Platform + Platform + Platform Build Tools + Platform Images + Platform Releng Aggregator + Platform Root + Platform UI + Platform Website + SWT + SWT Binaries + SWT Website), activate
...

Committing the provisioning operation.
ERROR: org.eclipse.pde.core code=0 Problems occurred while resolving the target contents
  at org.eclipse.oomph.util.OomphPlugin.coreException(OomphPlugin.java:296)
  at org.eclipse.oomph.util.pde.TargetPlatformUtil.activateTargetDefinition(TargetPlatformUtil.java:151)
  at org.eclipse.oomph.targlets.internal.core.TargletContainer.forceUpdate(TargletContainer.java:905)
  at org.eclipse.oomph.setup.targlets.impl.TargletTaskImpl$4.run(TargletTaskImpl.java:1232)
  at org.eclipse.oomph.util.pde.TargetPlatformUtil.runWithTargetPlatformService(TargetPlatformUtil.java:120)
  at org.eclipse.oomph.setup.targlets.impl.TargletTaskImpl.perform(TargletTaskImpl.java:1092)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.doPerformNeededSetupTasks(SetupTaskPerformer.java:3864)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer$WorkspaceUtil$1.run(SetupTaskPerformer.java:5200)
  at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2453)
  at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2478)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer$WorkspaceUtil.performNeededSetupTasks(SetupTaskPerformer.java:5193)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.performNeededSetupTasks(SetupTaskPerformer.java:3798)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.performTriggeredSetupTasks(SetupTaskPerformer.java:3773)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.perform(SetupTaskPerformer.java:3651)
  at org.eclipse.oomph.setup.ui.wizards.ProgressPage$9.run(ProgressPage.java:592)
  at org.eclipse.oomph.setup.ui.wizards.ProgressPage$11$1.run(ProgressPage.java:721)
  at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
  ERROR: org.eclipse.oomph.targlets.core code=0 Resolution problems
    ERROR: org.eclipse.oomph.targlets.core code=0 Problems resolving composed target 'eclipse-sdk-prereqs'
      ERROR: org.eclipse.pde.core code=0 Problems occurred while resolving the target contents
        ERROR: org.eclipse.pde.core code=0 Problems loading repositories
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit com.jcraft.jsch 0.1.55.v20230916-1400
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit com.jcraft.jsch.source 0.1.55.v20230916-1400
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.ant 1.10.14.v20230922-1200
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.ant.source 1.10.14.v20230922-1200
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.batik.css 1.17.0.v20231009-1000
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.batik.util 1.17.0.v20231009-1000
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.batik.i18n 1.17.0.v20231009-1000
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.batik.constants 1.17.0.v20231009-1000
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.batik.css.source 1.17.0.v20231009-1000
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.batik.util.source 1.17.0.v20231009-1000
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.batik.i18n.source 1.17.0.v20231009-1000
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.batik.constants.source 1.17.0.v20231009-1000
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.xmlgraphics 2.9.0.v20230916-1600
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.xmlgraphics.source 2.9.0.v20230916-1600
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.junit 4.13.2.v20230809-1000
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.junit.source 4.13.2.v20230809-1000
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.lucene.core 9.8.0.v20230929-1030
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.lucene.analysis-smartcn 9.8.0.v20230929-1030
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.lucene.analysis-common 9.8.0.v20230929-1030
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.lucene.core.source 9.8.0.v20230929-1030
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.lucene.analysis-smartcn.source 9.8.0.v20230929-1030
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.apache.lucene.analysis-common.source 9.8.0.v20230929-1030
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.hamcrest.core 2.2.0.v20230809-1000
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.hamcrest.core.source 2.2.0.v20230809-1000
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.jdom 1.1.3.v20230812-1600
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.jdom.source 1.1.3.v20230812-1600
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit com.sun.jna 5.13.0.v20230812-1000
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.eclipse.orbit.xml-apis-ext 1.0.0.v20230923-0644
          ERROR: org.eclipse.pde.core code=0 Unable to locate installable unit org.eclipse.orbit.xml-apis-ext.source 1.0.0.v20230923-0644
```

So I opened the file `eclipse-sdk-prereqs.target` present in my WS and got this error instead:

![grafik](https://github.com/eclipse-equinox/p2/assets/2205684/f397a32d-f512-40e5-9354-a99707c1bbbd)

Looking closer, this is the stack trace from the event in the **Error Log**
```java
java.lang.NoSuchMethodError: 'org.eclipse.equinox.p2.query.IQueryable org.eclipse.equinox.internal.p2.director.PermissiveSlicer.slice(java.util.Collection, org.eclipse.core.runtime.IProgressMonitor)'
	at org.eclipse.pde.internal.core.target.P2TargetUtils.slice(P2TargetUtils.java:1364)
	at org.eclipse.pde.internal.core.target.P2TargetUtils.resolveWithSlicer(P2TargetUtils.java:1284)
	at org.eclipse.pde.internal.core.target.P2TargetUtils.synchronize(P2TargetUtils.java:831)
	at org.eclipse.pde.internal.core.target.TargetDefinition.resolve(TargetDefinition.java:404)
	at org.eclipse.pde.internal.ui.editor.targetdefinition.TargetEditor$TargetChangedListener$1.run(TargetEditor.java:651)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```
I noticed that the missing method is actually the **new** version of `slice` _i.e._ the one with the `Collection` in the parameters, which is odd.

Any idea how to fix this? 

## About the Oomph Installer
* Version: **1.26.0 Build 5632**
<img width="655" alt="grafik" src="https://github.com/eclipse-equinox/p2/assets/2205684/5db93aa3-00b6-4855-b98e-5076318781ac">

Reply:
The missing new methods does looks odd indeed.  At this point it looks like you would have to run the p2 task to update the installation.    Another workaround is to delete this part of the task (the one that composed the actual *.target file) temporarily:

![image](https://github.com/eclipse-equinox/p2/assets/208716/7fbde859-b8ec-4107-adfc-2e46375638b0)

@HannesWell 

I've lost count of how many times the resolution of the Platform's real *.target causes problem.  It's certainly very good that one can get any target platform updates immediately rather than wait for the next I-Build, but very frustrating how often PDE fails to resolve it.  Perhaps I need to do something in Oomph to be able to ignore failures...

Reply:
Just for the record PDE was fixed with https://github.com/eclipse-pde/eclipse.pde/pull/808.



> Any idea how to fix this?

From the exception and the screenshot of the installed product I assume your installation has not the latest (I-Build) version of the `org.eclipse.equinox.p2.director` installed but somehow has the latest (I-build) version of `o.e.pde.core`?
Can you provide the exact version (including qualifier) of both bundles from the 'About' dialog?

Reply:
> I've lost count of how many times the resolution of the Platform's real *.target causes problem. It's certainly very good that one can get any target platform updates immediately rather than wait for the next I-Build, but very frustrating how often PDE fails to resolve it. Perhaps I need to do something in Oomph to be able to ignore failures...

Sadly I noticed that as well.
I could solve the resolution errors if I manually opended the prerequ-sdk target in the editor and after the resolution completed run the Oomph-Setup again.
Maybe the equivalent can be done in Oomph automatically? I can look-up the code or create a PR for Oomph if you like.

In general I already started to look into making the target-platform resolution more stable in PDE and preventing the quirks of unncessary re-resolved profiles, I already found a few things that are very likely causing problems but I'm currently busy with other things (i.e. the Windows-Defender auto-fix) so it might take a bit until I can complete these fixes.

Reply:
Yes, allowing it to resolve in the open editor is the workaround that I use too...

This is the location where we resolve the named target platform:

https://github.com/eclipse-oomph/oomph/blob/7a6ea7b978ff69cd6f94f5b8b446f71a825502c3/plugins/org.eclipse.oomph.targlets.core/src/org/eclipse/oomph/targlets/internal/core/TargletContainer.java#L1416

There is no other API on ITargetDefinition for resolving...

Reply:
> There is no other API on ITargetDefinition for resolving...

Even though I don't know in particular what the detailed problem is, but usually you can increment the SequenceNumber (even though there seems to be no API for that so you need to check/cast for `TargetDefinition` implementation).

Beside that, the pattern to force a (re)resolve is to call `ITargetplatform#setLocations(...)` followed by `ITargetplatform.resolve(...)`

Reply:
Sorry for the late response @HannesWell , I just got in front of the computer again

>  Can you provide the exact version (including qualifier) of both bundles from the 'About' dialog?

org.eclipse.equinox.p2.director: 2.6.200.v20230914-2056
org.eclipse.pde.core: 3.17.200.v20231022-0753

Reply:
> org.eclipse.equinox.p2.director: 2.6.200.v20230914-2056\norg.eclipse.pde.core: 3.17.200.v20231022-0753

The qualifier in both versions should be relativly similar +-1 day, but as anticipated you only got PDE.core from the nightly build, but not p2.director.
No clue why this happens.

Reply:
The problem persist even after a 2nd and a 3rd attempt today, This is what I tried:

**2nd attempt**: I ran _Help > Perform setup tasks_ today --> only the version of the `o.e.pde.core` got upgraded to 3.17.200.v20231023-0618 (`o.e.equinox...` remained the same)

**3rd attempt**: I downloaded the newest Oomph installer (`1.31.0 Build 213`) and installed from scratch --> same result: `o.e.pde.core` got upgraded to 3.17.200.v20231023-0618  and `o.e.equinox...` remained the same.

How inconvenient, this effectively keeps me from contributing from my Mac :-\

Reply:
I ran the p2 director task yesterday and see this version installed:

![image](https://github.com/eclipse-equinox/p2/assets/208716/bce59d2c-c556-4814-b2ab-2b24642f234b)

The director task is installing from these sites:

![image](https://github.com/eclipse-equinox/p2/assets/208716/83e59ad5-b48b-49b1-ad45-0a027e1db9f3)

There is definitely a new version of the p2 director bundle in that repository.  Moreover, I have no idea which of those repositories could (does) provide an older version...

Can you show a screen grab of the p2 director task details?

Reply:
I also see https://download.eclipse.org/eclipse/updates/4.30-I-builds in the _P2 Director_ task when performing the setup tasks.
 
<img width="1145" alt="grafik" src="https://github.com/eclipse-equinox/p2/assets/2205684/26cbbdf7-5823-4b2f-8dcf-2f31f3ab5c6d">


And here is the full log after I perform the setup tasks: [fullLog.txt](https://github.com/eclipse-equinox/p2/files/13176891/fullLog.txt)

Reply:
Now it's clear what the problem is!

You did not really use the configuration:

https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md#creating-an-eclipse-development-environment

So you did not install the Eclipse SDK product as specified in the configuration:

![image](https://github.com/eclipse-equinox/p2/assets/208716/fd2d248c-3562-4a3f-ab1e-ba09d37f7ef9)

Instead you installed the Eclipse Committers product.  So you have a combination of content that's installed from the EPP/SimRel M1 along with content installed from the 4.30 I-Builds.  That's because the products have some very specific ranges  on things that are intended never to be uninstalled from such a product:

![image](https://github.com/eclipse-equinox/p2/assets/208716/9fa3e564-db6a-4242-a10d-e602add9d6b0)

If you're going to do things a bit more manually, be sure to use the Eclipse SDK product:

![image](https://github.com/eclipse-equinox/p2/assets/208716/5e0ac149-45e7-4655-931d-de2d1670243a)

Reply:
🤯 

This one is going directly to my favorites. 

Thanks @merks, that solved my problem! 🥇